TAScheduler/views_methods.py

```diff
@@ -193,28 +194,61 @@ def __init__(self, course_info):
         self.course_database = course_info
 
     def addInstructor(self, active_instr):
-        pass
+        if not isinstance(active_instr, InstructorObj):
+            raise TypeError("active_instr is not an instance of InstructorObj")
+        if InstructorToCourse.objects.filter(course=self.course_database).count() >= self.course_database.max_instructors:
+            raise ValueError("This course has reached the maximum number of instructors")
+        if InstructorToCourse.objects.filter(
+                instructor=active_instr.database).count() >= active_instr.database.max_assignments:
+            raise ValueError("Instructor has reached the maximum number of course assignments")
+        InstructorToCourse.objects.create(instructor=active_instr.database, course=self.course_database)
 
     def addTa(self, active_ta):
-        pass
+        if not isinstance(active_ta, TAObj):
+            raise TypeError("active_ta is not an instance of TAObj")
+        if TAToCourse.objects.filter(course=self.course_database).count() >= self.course_database.max_tas:
+            raise ValueError("This course has reached the maximum number of TAs")
+        if TAToCourse.objects.filter(ta=active_ta.database).count() >= active_ta.database.max_assignments:
+            raise ValueError("TA has reached the maximum number of course assignments")
+        TAToCourse.objects.create(ta=active_ta.database, course=self.course_database)
 
     def removeAssignment(self, active_user):
-        pass
+        # to implement a way to determine if a user is a TA or Instructor
+        if isinstance(active_user, InstructorObj):
+            InstructorToCourse.objects.filter(course=self.course_database, instructor=active_user.database).delete()
+        elif isinstance(active_user, TAObj):
+            TAToCourse.objects.filter(course=self.course_database, ta=active_user.database).delete()
+        else:
+            raise TypeError("The active_user must be an instance of InstructorObj or TAObj")
 
     def removeCourse(self):
-        pass
+        self.course_database.delete()
 
     def editCourse(self, course_info):
-        pass
+        for attr, value in course_info.items():
+            setattr(self.course_database, attr, value)
+        self.course_database.full_clean()  # call the model's clean() method to validate the fields.
+        self.course_database.save()
 
     def getAsgmtsForCrse(self):
-        pass
+        return {
```

Comment:
unless i am missing something, I am not sure that "self.course_database.instructortocourse_set" works as intended (course doesn't have a .instructortocourse_set field) ... for my "getAsmgts" I just did a objects.filter() for that objects and returned the query set (1 line of code) -> in your case, maybe do 2 .filters and combine them into a bigger queryset before returning?

Reply:
Looks good to me. A dictionary with two lists is just as easy to use and it separates them by user type which is nice to have